File: src/local_learning_node.py
#!/usr/bin/env python3
import os
from typing import Optional

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from global_to_polar_cpp.msg import PolarGrid
from f1tenth_planning_custom_msgs.msg import PathWithVelocity, Waypoint
from message_filters import Subscriber, ApproximateTimeSynchronizer
from ament_index_python.packages import get_package_share_directory
import torch


class LocalLearningPlanner(Node):
    def __init__(self) -> None:
        super().__init__('local_learning')

        package_dir = get_package_share_directory('local_learning')
        default_model_path = os.path.join(
            package_dir, 'model', 'mobilenet_trained_updated.pt'
        )
        self.declare_parameter('model_path', default_model_path)
        model_path = self.get_parameter('model_path').get_parameter_value().string_value
        if not os.path.isabs(model_path):
            model_path = os.path.join(package_dir, model_path)

        self.model: Optional[torch.jit.ScriptModule] = None
        if os.path.exists(model_path):
            try:
                self.model = torch.jit.load(model_path)
                self.model.eval()
            except Exception as e:  # noqa: BLE001
                self.get_logger().warn(f'Failed to load model: {e}')
                self.model = None
        else:
            self.get_logger().warn(
                f'Model file not found at {model_path}, planner will not run.'
            )

        self.scan_sub = Subscriber(self, LaserScan, '/scan')
        self.grid_sub = Subscriber(self, PolarGrid, '/polar_grid')
        self.sync = ApproximateTimeSynchronizer(
            [self.scan_sub, self.grid_sub], queue_size=10, slop=0.1
        )
        self.sync.registerCallback(self.synced_callback)

        self.publisher = self.create_publisher(
            PathWithVelocity, '/planned_path_with_velocity', 10
        )

    def preprocess(self, scan: LaserScan, grid: PolarGrid) -> Optional[torch.Tensor]:
        if len(scan.ranges) != 1080 or len(grid.ranges) != 1080:
            self.get_logger().warn(
                'Expected 1080 points in both scan and grid.'
            )
            return None

        scan_tensor = torch.tensor(scan.ranges, dtype=torch.float32)
        grid_tensor = torch.tensor(grid.ranges, dtype=torch.float32)
        scan_tensor = torch.nan_to_num(scan_tensor, nan=0.0, posinf=0.0, neginf=0.0)
        grid_tensor = torch.nan_to_num(grid_tensor, nan=0.0, posinf=0.0, neginf=0.0)
        stacked = torch.stack((scan_tensor, grid_tensor)).unsqueeze(0)
        return stacked

    def synced_callback(self, scan: LaserScan, grid: PolarGrid) -> None:
        if self.model is None:
            return

        input_tensor = self.preprocess(scan, grid)
        if input_tensor is None:
            return

        with torch.no_grad():
            output = self.model(input_tensor).squeeze(0).cpu()

        msg = PathWithVelocity()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'map'

        for row in output:
            wp = Waypoint()
            wp.x = float(row[0])
            wp.y = float(row[1])
            wp.yaw = float(row[2])
            wp.velocity = float(row[3])
            msg.points.append(wp)

        while len(msg.points) < 10:
            msg.points.append(Waypoint())

        self.publisher.publish(msg)


def main(args: Optional[list[str]] = None) -> None:
    rclpy.init(args=args)
    node = LocalLearningPlanner()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/local_learning_node.cpp
#include <memory>
#include <optional>
#include <filesystem>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "global_to_polar_cpp/msg/polar_grid.hpp"
#include "f1tenth_planning_custom_msgs/msg/path_with_velocity.hpp"
#include "f1tenth_planning_custom_msgs/msg/waypoint.hpp"

#include "message_filters/subscriber.h"
#include "message_filters/synchronizer.h"
#include "message_filters/sync_policies/approximate_time.h"

#include "ament_index_cpp/get_package_share_directory.hpp"

#include <torch/torch.h>
#include <torch/script.h>

class LocalLearningPlanner : public rclcpp::Node
{
public:
  using Scan = sensor_msgs::msg::LaserScan;
  using Grid = global_to_polar_cpp::msg::PolarGrid;
  using PathWithVelocity = f1tenth_planning_custom_msgs::msg::PathWithVelocity;
  using Waypoint = f1tenth_planning_custom_msgs::msg::Waypoint;

  LocalLearningPlanner()
  : Node("local_learning"),
    scan_sub_(this, "/scan"),
    grid_sub_(this, "/polar_grid"),
    sync_(SyncPolicy(10), scan_sub_, grid_sub_)
  {
    auto package_dir = ament_index_cpp::get_package_share_directory("local_learning");
    std::string default_model_path = package_dir + "/model/mobilenet_trained_updated.pt";
    this->declare_parameter("model_path", default_model_path);
    std::string model_path = this->get_parameter("model_path").as_string();
    if (!std::filesystem::path(model_path).is_absolute()) {
      model_path = package_dir + std::string("/") + model_path;
    }

    if (std::filesystem::exists(model_path)) {
      try {
        model_ = std::make_shared<torch::jit::script::Module>(torch::jit::load(model_path));
        model_->eval();
      } catch (const c10::Error &e) {
        RCLCPP_WARN(this->get_logger(), "Failed to load model: %s", e.what());
        model_.reset();
      }
    } else {
      RCLCPP_WARN(this->get_logger(),
                  "Model file not found at %s, planner will not run.",
                  model_path.c_str());
    }

    sync_.registerCallback(std::bind(&LocalLearningPlanner::synced_callback,
                                     this,
                                     std::placeholders::_1,
                                     std::placeholders::_2));

    publisher_ = this->create_publisher<PathWithVelocity>(
        "/planned_path_with_velocity", 10);
  }

private:
  using SyncPolicy = message_filters::sync_policies::ApproximateTime<Scan, Grid>;

  std::optional<torch::Tensor> preprocess(const Scan &scan, const Grid &grid)
  {
    if (scan.ranges.size() != 1080 || grid.ranges.size() != 1080) {
      RCLCPP_WARN(this->get_logger(),
                  "Expected 1080 points in both scan and grid.");
      return std::nullopt;
    }
    auto scan_tensor = torch::from_blob(
        const_cast<float *>(scan.ranges.data()), {1080}, torch::kFloat32).clone();
    auto grid_tensor = torch::from_blob(
        const_cast<float *>(grid.ranges.data()), {1080}, torch::kFloat32).clone();

    scan_tensor = torch::nan_to_num(scan_tensor, 0.0, 0.0, 0.0);
    grid_tensor = torch::nan_to_num(grid_tensor, 0.0, 0.0, 0.0);

    auto stacked = torch::stack({scan_tensor, grid_tensor}).unsqueeze(0);
    return stacked;
  }

  void synced_callback(const Scan::ConstSharedPtr scan,
                       const Grid::ConstSharedPtr grid)
  {
    if (!model_) {
      return;
    }
    auto input_opt = preprocess(*scan, *grid);
    if (!input_opt) {
      return;
    }

    torch::NoGradGuard no_grad;
    auto output = model_->forward({*input_opt}).toTensor();
    output = output.squeeze(0).to(torch::kCPU);

    PathWithVelocity msg;
    msg.header.stamp = this->now();
    msg.header.frame_id = "map";

    auto accessor = output.accessor<float, 2>();
    for (int i = 0; i < output.size(0); ++i) {
      Waypoint wp;
      wp.x = accessor[i][0];
      wp.y = accessor[i][1];
      wp.yaw = accessor[i][2];
      wp.velocity = accessor[i][3];
      msg.points.push_back(wp);
    }

    while (msg.points.size() < 10) {
      msg.points.emplace_back();
    }

    publisher_->publish(msg);
  }

  message_filters::Subscriber<Scan> scan_sub_;
  message_filters::Subscriber<Grid> grid_sub_;
  message_filters::Synchronizer<SyncPolicy> sync_;
  rclcpp::Publisher<PathWithVelocity>::SharedPtr publisher_;
  std::shared_ptr<torch::jit::script::Module> model_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<LocalLearningPlanner>());
  rclcpp::shutdown();
  return 0;
}

